### PR TITLE
History

### PR DESCRIPTION
**Pull request by scottjehl**
@johnbender

### CODE REVIEW

**Comment by johnbender** [comment]
@scottjehl 

I will have this reviewed by tomorrow at 1pm!

File: dist/dialog.build.js [hunk 1]
```diff
@@ -24,10 +24,19 @@ window.jQuery = window.jQuery || window.shoestring;
...
+
+		// when dialog first inits, save a reference to the initial hash so we can know whether
+		// there's room in the history stack to go back or not when closing
+		this.initialLocationHash = w.location.hash;
+
+		// the dialog's url hash is different from the dialog's actual ID attribute
+		// this is because pairing the ID directly makes the browser jump to the top of the dialog,
+		// rather than allowing us to space it off the top of the viewport.
+		// also, if the dialog has a data-nohistory attr, this property will prevent its findability for onload and hashchanges
+		this.nohistory = this.$el.is( '[data-dialog-nohistory]' );
```
Comment (johnbender) [inline dist/dialog.build.js]: Nit: I think the convention we use (and others use) is `is<boolean phrase>` or `has<noun phrase>` so in this has it would be `hasHistory`?

File: dist/dialog.linker.build.js [hunk 2]
```diff
@@ -71,17 +78,21 @@ window.jQuery = window.jQuery || window.shoestring;
 		this.$background.attr( "id", this.$el.attr( "id" ) + "-background" );
 		this._setBackgroundTransparency();
 
-		if( this.isSetScrollPosition() ) {
-			this.scroll = "pageYOffset" in w ? w.pageYOffset : ( docElem.scrollY || docElem.scrollTop || ( body && body.scrollY ) || 0 );
-			this.$el[ 0 ].style.top = this.scroll + "px";
-		} else {
-			this.$el[ 0 ].style.top = '';
-		}
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Is removing this related to the history change? 

It looks like we removed a feature, i.e. `data-set-position-media`. That might make it harder for projects update to this new version (which will likely be major).

File: dist/dialog.linker.build.js [hunk 3]
```diff
@@ -71,17 +78,21 @@ window.jQuery = window.jQuery || window.shoestring;
...
 
 		$html.addClass( cl.open );
 		this.isOpen = true;
 
-		window.location.hash = this.hash;
+		var cleanHash = w.location.hash.replace( /^#/, "" );
+		var lastHash = w.location.hash.split( "#" ).pop();
+
+		if( cleanHash.indexOf( "-dialog" ) > -1 && lastHash !== this.hash ){
+			w.location.hash += "#" + this.hash;
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Just checking my understanding, if a dialog is open at `http://foo#bar-dialog` and a nested dialog is opened then it will be `http://foo#bar-dialog#baz-dialog`. Is there any concern about weird behavior when we concatenate hashes that way?

File: dist/dialog.linker.build.js [hunk 4]
```diff
@@ -71,17 +78,21 @@ window.jQuery = window.jQuery || window.shoestring;
...
+		}
+		else if( lastHash !== this.hash ) {
+			w.location.hash = this.hash;
+		}
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Nit: can we just append the hash whenever it's not the same as the `lastHash`? Then we don't have to worry about the corner case where `-dialog` somehow ends up in the hash for other reasons.

That is, it seems like the fact that `-dialog` is in the hash shouldn't matter?

File: dist/dialog.linker.build.js [hunk 5]
```diff
@@ -229,26 +263,35 @@ window.jQuery = window.jQuery || window.shoestring;
...
+
+        // if the hash matches this dialog's, open!
+        if( hash === dialog.hash ){
+          if( !dialog.nohistory ){
+            dialog.open();
+          }
+        }
+        // if it doesn't match...
+				else {
+          dialog.close();
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: This hunk has whitespace issues.

File: dist/dialog.linker.build.js [hunk 6]
```diff
@@ -229,26 +263,35 @@ window.jQuery = window.jQuery || window.shoestring;
...
 				}
 			});
 
-			// open on matching a[href=#id] click
+      // open on matching a[href=#id] click
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Whitespace

File: dist/dialog.linker.build.js [hunk 7]
```diff
@@ -229,26 +263,35 @@ window.jQuery = window.jQuery || window.shoestring;
...
 			$( doc ).bind( "click", function( e ){
 				var $matchingDialog, $a;
 
 				$a = $( e.target ).closest( "a" );
 
+
 				if( !dialog.isOpen && $a.length && $a.attr( "href" ) ){
+          var id = $a.attr( "href" ).replace( /^#/, "" );
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Whitespace

File: dist/dialog.linker.build.js [hunk 8]
```diff
@@ -271,8 +316,9 @@ window.jQuery = window.jQuery || window.shoestring;
...
 	};
 
-	// auto-init
-	$(function(){
-		$( ".dialog" ).dialog();
+  // auto-init on enhance
+	$( w.document ).bind( "enhance", function( e ){
+    var target = e.target === w.document ? "" : e.target;
+		$( "." + pluginName, e.target ).add( target ).filter( "." + pluginName )[ pluginName ]();
 	});
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Whitespace.

File: dist/dialog.linker.build.js [hunk 9]
```diff
@@ -102,13 +113,24 @@ window.jQuery = window.jQuery || window.shoestring;
...
+		// then we need to change the hash to remove it, either by going back if we can, or by adding a history
+		// state that doesn't have it at the end
+		if( window.location.hash.split( "#" ).pop() === this.hash ){
+			// let's check if the first segment in the hash is the same as the first segment in the initial hash
+			// if not, it's safe to use back() to close this out and clean the hash up
+			var firstHashSegment = window.location.hash.split( "#" )[ 1 ];
+			var firstInitialHashSegment = this.initialLocationHash.split( "#" )[ 1 ];
+			if( firstHashSegment && firstInitialHashSegment && firstInitialHashSegment !== firstHashSegment ){
+				window.history.back();
+			}
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: I think if we end up changing the hash to something other than the initial hash value and then change it back to the initial hash value we would still want to go back. Example:

1. initial hash: `#foo-dialog`
2. close foo dialog: empty hash
3. open foo dialog: `#foo-dialog`
4. should use `window.history.back`

File: dist/dialog.linker.build.js [hunk 10]
```diff
@@ -102,13 +113,24 @@ window.jQuery = window.jQuery || window.shoestring;
...
+			// otherwise, if it's the same starting hash as it was at init time,
+			// we can't trigger back to close the dialog, as it might take us elsewhere.
+			// so we have to go forward and create a new hash that does not have this dialog's hash at the end
+			else {
+				var escapedRegexpHash = this.hash.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1");
+				window.location.hash = window.location.hash.replace( new RegExp( "#" + escapedRegexpHash + "$" ), "" );
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Why don't we just go to the empty hash when a dialog that is opened on page load gets closed? I think this covers a case I'm not considering.

File: dist/dialog.linker.build.js [hunk 11]
```diff
@@ -150,26 +170,24 @@ window.jQuery = window.jQuery || window.shoestring;
...
 				var dialogClasses = $a.attr( "data-dialog-addclass" ) || "";
+				var dialogNoHistory = $a.is( "[data-dialog-nohistory]" );
 				var id;
 
-				if( linkId ) {
-					id = linkId + "-dialog";
-				} else {
-					id = "dialog-" + new Date().getTime();
+				if( linkHref ) {
+					id = linkHref;
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: I guess previously we were looking at the anchor's id and then that was being used to target the dialog?

File: dist/dialog.linker.build.js [hunk 12]
```diff
@@ -150,26 +170,24 @@ window.jQuery = window.jQuery || window.shoestring;
...
 				}
 
 				$a
 					.attr("href", "#" + id )
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: It seems like if we're assuming the `id` value comes from the link's `href` attribute we don't need to set it again with an extra hash right?

File: dist/dialog.build.js (hunk 1 above)
Comment (scottjehl) [inline dist/dialog.build.js]: hmm. sure I'd be open to flipping that if you think we should make it look true as a default.

File: dist/dialog.linker.build.js (hunk 3 above)
Comment (scottjehl) [inline dist/dialog.linker.build.js]: correct.
Um, it seemed to be an allowed character in the hash across browsers I tested. I'm open to other characters or patterns that would denote a new segment. Thoughts?

File: dist/dialog.linker.build.js [hunk 13]
```diff
@@ -24,10 +24,19 @@ window.jQuery = window.jQuery || window.shoestring;
...
+
+		// when dialog first inits, save a reference to the initial hash so we can know whether
+		// there's room in the history stack to go back or not when closing
+		this.initialLocationHash = w.location.hash;
+
+		// the dialog's url hash is different from the dialog's actual ID attribute
+		// this is because pairing the ID directly makes the browser jump to the top of the dialog,
+		// rather than allowing us to space it off the top of the viewport.
+		// also, if the dialog has a data-nohistory attr, this property will prevent its findability for onload and hashchanges
+		this.nohistory = this.$el.is( '[data-dialog-nohistory]' );
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Nit: I think the convention we (and others) normally is: `is<adjective>` or `has<noun>`. So in this case it would be `this.hasHistory`.

File: dist/dialog.linker.build.js [hunk 14]
```diff
@@ -198,14 +216,30 @@ window.jQuery = window.jQuery || window.shoestring;
...
+	$( w ).bind( "hashchange load", function(){
+		var hash = w.location.hash.split( "#" ).pop();
+		var id = hash.replace( /-dialog$/, "" );
+		var $ajaxLink = $( 'a[href="' + id +'"][data-dialog-link]' );
+		// if the link specified nohistory, don't click it
+		var nohistory = $ajaxLink.is( "[data-dialog-nohistory]" );
+		var $dialogInPage = $( '.dialog[id="' + id + '"]' );
+		if( $ajaxLink.length && !nohistory && !$dialogInPage.length ){
+			$ajaxLink.eq( 0 ).trigger( "click" );
+		}
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: Shouldn't we open the dialog through some more direct means? Why does this only apply to ajax content dialogs?

File: dist/dialog.linker.build.js (hunk 4 above)
Comment (scottjehl) [inline dist/dialog.linker.build.js]: so, the idea here is to only add to the hash if it's already a dialog hash. Adding to hashes that reference something else, say... an odd element's ID in the page, seems odd to me, since that prior hash is unrelated to dialog and doesn't help anything being in the dialog hash.

File: dist/dialog.linker.build.js (hunk 9 above)
Comment (scottjehl) [inline dist/dialog.linker.build.js]: Hmm. I don't think actively opening a dialog should ever trigger `history.back`. Opening feels like a step forward, and additive to history, even if it's a place you've been before. 

This logic is just about what to do when closing, which is destructive to history if we are able to use it that way... its just that sometimes we can't because the open dialog is the first thing in our session.

File: dist/dialog.linker.build.js (hunk 10 above)
Comment (scottjehl) [inline dist/dialog.linker.build.js]: So this covers the case of changing the hash to a step back in the hash path on close. Basically, it allows us to send a nested dialog path that actually respects its parent context upon closing, by changing the hash to each respective parent until it's back to the page.

File: dist/dialog.linker.build.js (hunk 11 above)
Comment (scottjehl) [inline dist/dialog.linker.build.js]: Yes. Which seemed an odd requirement for one, but it gave us a "unique" hash to work with once the dialog was created. Still, this seems more direct I think, using the page source.

Thoughts?

File: dist/dialog.linker.build.js (hunk 5 above)
Comment (scottjehl) [inline dist/dialog.linker.build.js]: hmm yeah

File: dist/dialog.linker.build.js (hunk 3 above)
Comment (johnbender) [inline dist/dialog.linker.build.js]: Nope, seems good to me at the reading of the hash is clear in that case. 👍

File: dist/dialog.linker.build.js (hunk 9 above)
Comment (johnbender) [inline dist/dialog.linker.build.js]: Sorry, I borked the list I wrote, second try:

1. initial hash: #foo-dialog
2. close foo dialog: empty hash
3. open foo dialog: #foo-dialog
4. close foo dialog: should use window.history.back

File: dist/dialog.linker.build.js (hunk 14 above)
Comment (scottjehl) [inline dist/dialog.linker.build.js]: So John this ONLY covers the case when the hash matches an ajax dialog link that has not yet been clicked, and thus, has no dialog to open yet. It'll only happen once because once the dialog is created, that `$dialogInPage` var will be true and prevent this behavior from applying the click. 

After it's created, it'll still be opened by our ordinary dialog hashchange listener.

File: dist/dialog.linker.build.js (hunk 9 above)
Comment (scottjehl) [inline dist/dialog.linker.build.js]: ah, ok. You're right. So... maybe all we need to know is that the hash has changed at least once before using back? Hmm... actually I guess at a minimum we'd need a `+/-` of hash states to track if we're beyond position 1...

File: dist/dialog.linker.build.js (hunk 10 above)
Comment (johnbender) [inline dist/dialog.linker.build.js]: I'm confused. The comment says "we can't trigger back to close the dialog" because it might take us away from the page if the page was opened with a dialog hash. So it seems like the idea here is to create a new hash to add a history entry so that if we close the dialog and back from that hash we'll just go back to the page as it was loaded. 

I'm proposing that we handle this in the `close` method. If `close` is called for a dialog that was opened by the initial hash then it should clear that hash which will add an entry to history. This operates under the assumption that we would open the dialog on going back from there if necessary which may not be true/correct.

File: dist/dialog.linker.build.js (hunk 11 above)
Comment (johnbender) [inline dist/dialog.linker.build.js]: This seems much more sane 👍

File: dist/dialog.linker.build.js (hunk 14 above)
Comment (johnbender) [inline dist/dialog.linker.build.js]: Yah I guess I was thinking that the dialog linker should extend the `open` method to create the dialog before opening it so that we can just call `open` and it will work in both cases.

**Comment by toddparker** [comment]
Issue to check:
If you have a dialog with no history open, then refresh/deep link to that page it won't re-open it. If you then click the link to re-open the same dialog it will open, but will going Back close it or navigate you back to the previous page?

File: dist/dialog.linker.build.js [hunk 15]
```diff
@@ -198,16 +243,30 @@ window.jQuery = window.jQuery || window.shoestring;
...
+  var Dialog = w.componentNamespace.Dialog,
+      doc = w.document,
+      pluginName = "dialog";
 
-	$.fn.dialog = function(){
+	$.fn[ pluginName ] = function(){
 		return this.each(function(){
-			var $el = $( this ), dialog = new Dialog( this );
-
-			$el.data( "instance", dialog );
```
Comment (johnbender) [inline dist/dialog.linker.build.js]: We need this back for a client project.
